### PR TITLE
[FLINK-15221][table api]support fault-tolerant semantic for kafka table api

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Currently users still can't use 'exactly-once' semantic in kafka table api. Therefore, we expose key `sink.semantic` and offer values `['exactly-once', 'at-least-once', 'none']` for users to use.*

## Brief change log

  - *Add option in KafkaOptions*
  - *Add validation for value of `sink.semantic`*
  - *Use parsed value of `sink.semantic` to build KafkaProducer*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests for new option `sink.semantic`*
  - *Added test when input invalid value for option `sink.semantic`*
 
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by fsk119** [comment]
@flinkbot run azure

**Comment by fsk119** [comment]
CC @wuchong

File: flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/table/Kafka011DynamicSink.java [hunk 1]
```diff
@@ -42,26 +44,31 @@ public Kafka011DynamicSink(
...
+			Optional<FlinkKafkaPartitioner<RowData>> partitioner,
+			String semantic) {
 		return new FlinkKafkaProducer011<>(
 				topic,
-				serializationSchema,
+				new KeyedSerializationSchemaWrapper<>(serializationSchema),
 				properties,
-				partitioner);
+				partitioner,
+				getSemantic(semantic),
```
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/table/Kafka011DynamicSink.java]: I have already moved getSemantic logic from KafkaDynamicSink. However, I just add  a method named `transformSemantic` to resolve the gap between KafaProducer.Semantic and value of 'sink.semantic'. This method will 
transform input of 'sink.semantic' and replace '-' with '_';

File: flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/table/Kafka010DynamicTableFactory.java [hunk 2]
```diff
@@ -65,18 +74,31 @@ protected KafkaDynamicSinkBase createKafkaTableSink(
...
+
+	@Override
+	public Set<ConfigOption<?>> optionalOptions() {
+		final Set<ConfigOption<?>> options = new HashSet<>();
+		options.add(PROPS_GROUP_ID);
+		options.add(SCAN_STARTUP_MODE);
+		options.add(SCAN_STARTUP_SPECIFIC_OFFSETS);
+		options.add(SCAN_STARTUP_TIMESTAMP_MILLIS);
+		options.add(SINK_PARTITIONER);
+		return options;
```
Comment (wuchong) [inline flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/table/Kafka010DynamicTableFactory.java:102]: remove sink semantic option.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaOptions.java [hunk 3]
```diff
@@ -222,8 +226,17 @@ private static void validateSinkSemantic(ReadableConfig tableOptions) {
 	// Utilities
 	// --------------------------------------------------------------------------------------------
 
-	public static String transformSemantic(String semantic){
-		return semantic.toUpperCase().replace('-', '_');
+	public static KafkaSemantic getSinkSemantic(String semantic){
```
Comment (wuchong) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaOptions.java:229]: `SinkSemantic` or `KafkaSinkSemantic`? It maybe confusing that it also works for source.

File: docs/dev/table/connectors/kafka.md [hunk 4]
```diff
@@ -207,6 +214,19 @@ However, it will cause a lot of network connections between all the Flink instan
 
 By default, a Kafka sink ingests data with at-least-once guarantees into a Kafka topic if the query is executed with [checkpointing enabled]({% link dev/stream/state/checkpointing.md %}#enabling-and-configuring-checkpointing).
 
+With Flink's checkpointing enabled, the `kafka` and `kafka-0.11` connectors can provide exactly-once delivery guarantees.
+
+Besides enabling Flink's checkpointing, you can also choose three different modes of operating chosen by passing appropriate `sink.semantic` option:
+
+ * `NONE`: Flink will not guarantee anything. Produced records can be lost or they can be duplicated.
```
Comment (wuchong) [inline docs/dev/table/connectors/kafka.md:221]:
```suggestion
 * `none`: Flink will not guarantee anything. Produced records can be lost or they can be duplicated.
```

File: docs/dev/table/connectors/kafka.md [hunk 5]
```diff
@@ -207,6 +214,19 @@ However, it will cause a lot of network connections between all the Flink instan
...
+ * `AT_LEAST_ONCE` (default setting): This guarantees that no records will be lost (although they can be duplicated).
```
Comment (wuchong) [inline docs/dev/table/connectors/kafka.md:222]:
```suggestion
 * `at-least-once` (default setting): This guarantees that no records will be lost (although they can be duplicated).
```

File: docs/dev/table/connectors/kafka.md [hunk 6]
```diff
@@ -207,6 +214,19 @@ However, it will cause a lot of network connections between all the Flink instan
...
+ * `EXACTLY_ONCE`: Kafka transactions will be used to provide exactly-once semantic. Whenever you write
```
Comment (wuchong) [inline docs/dev/table/connectors/kafka.md:223]:
```suggestion
 * `exactly-once`: Kafka transactions will be used to provide exactly-once semantic. Whenever you write
```

File: flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/table/Kafka010DynamicTableFactory.java [hunk 7]
```diff
@@ -65,18 +69,27 @@ protected KafkaDynamicSinkBase createKafkaTableSink(
...
 
 	@Override
 	public String factoryIdentifier() {
 		return IDENTIFIER;
 	}
+
+	@Override
+	public Set<ConfigOption<?>> optionalOptions() {
+		final Set<ConfigOption<?>> options = super.optionalOptions();
+		options.remove(SINK_SEMANTIC);
```
Comment (wuchong) [inline flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/table/Kafka010DynamicTableFactory.java:92]: Add a comment on this to explain why we remove sink semantic in 0.10

**Pull request closed without merging**